Ical.Net/TimeZone/VTimeZoneExtensions.cs

```diff
@@ -0,0 +1,150 @@
+﻿//
+// Copyright ical.net project maintainers and contributors.
+// Licensed under the MIT license.
+//
+
+#nullable enable
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Ical.Net.CalendarComponents;
+using Ical.Net.DataTypes;
+using Ical.Net.Evaluation;
+using NodaTime;
+using NodaTime.TimeZones;
+
+namespace Ical.Net.TimeZone;
+
+/// <summary>
+/// Extensions for the <see cref="VTimeZone"/> class.
+/// </summary>
+internal static class VTimeZoneExtensions
+{
+    /// <summary>
+    /// Converts a collection of <see cref="VTimeZone"/>s
+    /// to a collection of <see cref="CalDateTimeZone"/>s.
+    /// </summary>
+    /// <param name="vTimeZones"></param>
+    /// <returns></returns>
+    public static IEnumerable<CalDateTimeZone> ToDateTimeZones(this ICollection<VTimeZone> vTimeZones, int untilYear)
+        => vTimeZones.Select(vTimeZone => vTimeZone.ToDateTimeZone(untilYear));
+
+    /// <summary>
+    /// Converts a <see cref="VTimeZone"/> to a <see cref="CalDateTimeZone"/>.
+    /// </summary>
+    /// <exception cref="ArgumentException"></exception>
+    public static CalDateTimeZone ToDateTimeZone(this VTimeZone vTimeZone, int untilYear)
+    {
+        if (TryCreateIntervals(vTimeZone, untilYear, out var tzId, out var intervals))
+            return new CalDateTimeZone(tzId, intervals);
+
+        throw new ArgumentException($"VTimeZone '{vTimeZone.TzId}' does not contain valid TimeZoneInfos.", nameof(vTimeZone));
+    }
+
+    /// <summary>
+    /// Converts a <see cref="T:System.DateTime" /> of any kind to a <see cref="LocalDateTime"/> in the ISO calendar.
+    /// Then, the given <paramref name="offset"/> is applied to the <see cref="LocalDateTime"/> to create an <see cref="Instant"/>.
+    /// <para/>
+    /// Any <see cref="P:System.DateTime.Kind" /> will create the same day/hour/minute etc.
+    /// </summary>
+    private static Instant CreateInstant(DateTime dt, Offset offset)
+    {
+        return LocalDateTime
+            .FromDateTime(dt)
+            .WithOffset(offset)
+            .ToInstant();
+    }
+
+    /// <summary>
+    /// A temporary class to hold transition information
+    /// that can be used to create <see cref="ZoneInterval"/>s.
+    /// </summary>
+    private sealed class Transition(string tzName, Instant start, Offset offset, Offset savings, Instant? end = null)
+    {
+        public string TzName { get; } = tzName;
+        public Instant Start { get; } = start;
+
+        /// <summary>
+        /// Will be set in <see cref="CreateIntervalsFromTransitions"/>
+        /// </summary>
+        public Instant? End { get; set; } = end;
+        public Offset Offset { get; } = offset;
+        public Offset Savings { get; } = savings;
+    }
+
+    private static bool TryCreateIntervals(VTimeZone vTimeZone, int untilYear, out string tzId,
```

Comment:
Try to avoid the need for `untilYear` altogether? Alternatively just combine the individual IEnumerables to a single one using `.OrderedMerge()` and store the corresponding `IEnumerator`. When more data is needed, just continue the enumeration?